src/librustc_codegen_llvm/debuginfo/source_loc.rs

```diff
@@ -43,18 +45,43 @@ pub fn set_source_location<D>(
 
 #[derive(Copy, Clone, PartialEq)]
 pub enum InternalDebugLocation<'ll> {
-    KnownLocation { scope: &'ll DIScope, line: usize, col: usize },
+    KnownLocation {
+        scope: &'ll DIScope,
+        line: usize,
+        col: Option<NonZeroUsize>,
+    },
     UnknownLocation
 }
 
 impl InternalDebugLocation<'ll> {
-    pub fn new(scope: &'ll DIScope, line: usize, col: usize) -> Self {
+    pub fn new(scope: &'ll DIScope, line: usize, col: Option<NonZeroUsize>) -> Self {
         KnownLocation {
             scope,
             line,
             col,
         }
     }
+
+    pub fn from_span(cx: &CodegenCx<'ll, '_>, scope: &'ll DIScope, span: Span) -> Self {
+        let pos = cx.sess().source_map().lookup_char_pos(span.lo());
+
+        // FIXME: Rust likes to emit zero-width spans just after the end of a function. For now,
+        // zero-width spans to the preceding column to avoid emitting a column that points past the
+        // end of a line. E.g.,
+        //   |xyz = None
+        //   x|yz = 1
+        //   xy|z = 2
+        //
+        // See discussion in https://github.com/rust-lang/rust/issues/65437
```

Comment:
I don't really understand what this comment means. Maybe we can try to make it clearer to the casual reader.

Reply:
How's this? Or is it the table specifically that's confusing?

```rust       
// Rust likes to emit zero-width spans that point just after a closing brace to denote e.g.
// implicit return from a function. Instead of mapping zero-width spans to the column at
// `span.lo` like we do normally, map them to the column immediately before the span. This
// ensures that we point to a closing brace in the common case, and that debuginfo doesn't
// point past the end of a line. A zero-width span at the very start of the line gets
// mapped to `0`, which is used to represent "no column information" in DWARF. For example,
//
//   Span len = 0           Span len = 1
//
//   |xyz => 0 (None)        |x|yz => 1
//   x|yz => 1               x|y|z => 2
//
// See discussion in https://github.com/rust-lang/rust/issues/65437 for more info.
```

Reply:
Is there any documentation on when zero-width spans are emitted and for what reason? Or is this reverse-engineering the current (unspecified) behavior?

Reply:
This is the code responsible: https://github.com/rust-lang/rust/blob/ac45719c5d13cdcd6e9adc3f7d5807481c3260f3/src/librustc_mir/build/mod.rs#L633